docs/design/operator.md

```diff
@@ -0,0 +1,200 @@
+not be strictly limited by the extents of current-day technologies but should
+not dismiss the costs of waiting for and executing forward development in those
+technologies.
+
+In addition, this document only specifies the desired functionality for
+declaring a 1.0 release. Further enhancements and feature roadmaps will be
+captured elsewhere.
+
+### Motivation
+
+The Gluster operator aims to further streamline the experience of deploying and
+maintaining Gluster clusters for use in a Kubernetes environment. It will
+achieve this through the use of common Kubernetes components and paradigms,
+helping further integrate persistent storage into the Kubernetes experience.
+
+## Configurations
+
+The Gluster operator will be capable of managing a number of configurations of
+Gluster clusters. Gluster's flexibility allows it to run in many typs of
+environments and the operator will aim to preserve that flexibility as much as
+possible.
+
+### External Deployments
+
+External deployments are those in which the Gluster servers are not running as
+containers inside Kubernetes. They are running on external, potentially
+dedicated, storage nodes.
+
+### Hosted Deployments
+
+Hosted deployments are those in which the Gluster servers are running as
+containers but their storage devices are not managed by the operator. In these
+deployments, an administrator is required to maintain any underlying storage
+devices and inform the operator as to any changes in their configuration.
+
+### Managed Deployments
+
+Managed deployments are those in which the Gluster servers are running as
+containers and their storage devices are managed by the operator. These
+deployments are designed for cloud environments where storage devices can be
+provisioned programmatically.
+
+## Features
+
+Since the Gluster operator will support a variety of deployment configurations,
+it will not be feasible to provide identical feature sets across all
+configurations. The following sections note which features will be available to
+which configuration.
+
+All operational features will be designed to be as idempotent as possible. It is
+acknowledged that this may not be immediately possible due to restrictions in
+the underlying technology but an effort will still be made.
+
+### Deployment
+
+The Gluster operator will be able to deploy a new Gluster cluster using a set of
+configuration parameters. This configuration will allow for extensive
+customization if desired but will assume sane defaults for as many parameters as
+possible to allow for as minimal a configuration as possible.
+
+**Target Configurations:** Hosted, Managed
+
+### Import
+
+The Gluster operator will be able to import an existing Gluster cluster at any
+stage of initialization. The operator's desired configuration for the cluster
+will need to be compatible with the state of the cluster at the time of import.
+
+**Target Configurations:** External, Hosted, Managed
+
+#### Cluster Upgrade
```

Comment:
Can the cluster upgrade component be made dynamic enough to also be useful in other situations which require a rolling restart of the Gluster nodes? Effectively, it would handle bringing down Gluster cleanly, waiting for X and ensuring the health before continuing. Beyond the Gluster upgrade, the 2 primary examples of X being, upgrading the underlying nodes and restarting the VMs or Servers for OS Patching.

Reply:
The operator would not be managing the upgrade of any underlying infrastructure. In the case where a given node identity can not move between hosts, it would be up to the administrator to ensure they upgrade such hosts one at a time and wait for the Gluster cluster's health to restore. The only involvement of the operator in that case would be, at best, to report the state of health.

Reply:
Agreed, it would not handle those other tasks, but ideally, it would be able to aid them. Instead of everyone else needing to understand how to cleanly terminate gluster (daemon set / pod or otherwise) and check it's health when it comes back, it would be nice for other users and operators to be able to leverage the gluster operator for that functionality. If the tasks are broken out as discretely, that may be possible. Long and short, that kind of reusability is what I am hoping you take into consideration.

Reply:
At present, and best I can see, there's nothing more to bringing down a Gluster pod than any other pod. Kubernetes will take care to do as clean a shutdown as it can, to my understanding. As far as watching to health restoration, would simply exposing the state via a condition field in an object's Status be sufficient?

Reply:
> There is something more in that the volume with bricks hosted by that pod should be clean before shutting down. I'm not sure if sig lifecycle would have any opinions on how that could be ensured.

That's up to Gluster, not the operator. Kube sends a SIGTERM to the main process on pod delete, in our case some sort of init/systemd, which should take care of shutting things down gracefully.